### PR TITLE
RSDK-4495 Integration Tests

### PR DESCRIPTION
**Pull request by jeremyrhyde**
This PR adds complete integration testing for IMU. This includes large scale changes to the integration sensors, a refactoring of integration tests and updates to the sensorprocess to limit measurements in the following way:

- in offline mode, imu data is only added after the first lidar measurement has been added
- in offline mode, lidar data is only added if it is before the next imu measurement

it also introduces mutex locks to both the sensor process and integration sensors due to shared timing parameters.

New artifacts including a datafile for imu data has been added to this PR.

**JIRA Ticket:** https://viam.atlassian.net/browse/RSDK-4495

### CODE REVIEW

File: sensorprocess/sensorprocess_test.go [hunk 1]
```diff
@@ -595,6 +596,8 @@ func onlineModeIMUTestHelper(
 
 	// set lidar data rate to signify that we are in online mode
 	config.LidarDataRateMsec = 10
+	config.currentLidarData.time = time.Now().UTC().Add(-10 * time.Second)
```
Comment (jeremyrhyde) [inline sensorprocess/sensorprocess_test.go:599]: [note] `firstLidarReadingTime` and `currentLidarData.time` are used to determine when to allow imu data to be ingested. Both `currentLidarData.time` and `firstLidarReadingTime` must be before the imu time of `time.Now()`

File: sensors/sensors_test.go [hunk 2]
```diff
@@ -187,6 +187,8 @@ func TestTimedLidarSensorReading(t *testing.T) {
 
 	t.Run("when a live lidar succeeds, returns current time in UTC and the reading", func(t *testing.T) {
 		beforeReading := time.Now().UTC()
+		time.Sleep(10 * time.Millisecond)
```
Comment (jeremyrhyde) [inline sensors/sensors_test.go:190]: [note] wait was added as flaky tests would occur if execution is so fast that `tsr.ReadingTime `and `beforeReading` are equal

File: sensors/sensors_test.go [hunk 3]
```diff
@@ -205,7 +207,11 @@ func TestTimedLidarSensorReading(t *testing.T) {
 		tsr, err := goodReplayLidar.TimedLidarSensorReading(ctx)
 		test.That(t, err, test.ShouldBeNil)
 		test.That(t, tsr.Reading, test.ShouldNotBeNil)
-		test.That(t, tsr.ReadingTime, test.ShouldEqual, time.Date(2006, 1, 2, 15, 4, 5, 999900000, time.UTC))
+
+		readingTime, err := time.Parse(time.RFC3339Nano, s.TestTimestamp)
```
Comment (jeremyrhyde) [inline sensors/sensors_test.go:211]: [note] remove magic number and ensure equality by parsing/formatting the same way

File: sensorprocess/sensorprocess.go [hunk 4]
```diff
@@ -77,47 +77,79 @@ func (config *Config) StartLidar(
 
 // addLidarReading adds a lidar reading to the cartofacade.
 func (config *Config) addLidarReading(ctx context.Context) bool {
-	/*
-	 when the lidar data rate msec is non-zero, we assume the user wants to be in "online"
-	 mode and ensure the most recent scan gets processed by cartographer. If data rate msec
-	 is zero we process every scan in order
-	*/
+	// Process data in online or offline mode determined by the lidar's data rate
```
Comment (kim-mishra) [inline sensorprocess/sensorprocess.go:80]: Would it be possible to pull out some of the logic to get the next reading/timestamps. I think this function is getting increasingly large and sort of hard to follow

File: sensorprocess/sensorprocess.go [hunk 5]
```diff
@@ -202,44 +250,70 @@ func (config *Config) addIMUReading(
...
-		// sleep remainder of time interval
-		timeToSleep := tryAddIMUReading(ctx, sr, tsr.ReadingTime, *config)
-		time.Sleep(time.Duration(timeToSleep) * time.Millisecond)
+		// update stored imu time
+		config.Mutex.Lock()
+		config.currentIMUData.time = tsr.ReadingTime
+		config.currentIMUData.data = sr
+		currentLidarData := config.currentLidarData.time
+		cfg := *config
+		config.Mutex.Unlock()
```
Comment (kim-mishra) [inline sensorprocess/sensorprocess.go:259]: same comment as above on potentiall moving some of this logic into a helper func

File: sensorprocess/sensorprocess.go [hunk 6]
```diff
@@ -147,19 +179,37 @@ func tryAddLidarReadingUntilSuccess(ctx context.Context, reading []byte, reading
 
 // tryAddLidarReading adds a reading to the carto facade and does not retry (online).
 func tryAddLidarReading(ctx context.Context, reading []byte, readingTime time.Time, config Config) int {
-	startTime := time.Now()
+	startTime := time.Now().UTC()
```
Comment (kim-mishra) [inline sensorprocess/sensorprocess.go:182]: Does the replay sensor also provide UTC?

**Review by kim-mishra — COMMENTED** [review]
have some questions

File: sensorprocess/sensorprocess.go (hunk 6 above)
Comment (kim-mishra) [reply sensorprocess/sensorprocess.go:182]: just checked and saw that it did, was this causing issues?

File: integration_test.go [hunk 7]
```diff
@@ -30,6 +33,73 @@ import (
...
+		}
+	case runtime.GOOS == "linux" && useIMU:
+		expectedPos = r3.Vector{X: 3.2250269853115867, Y: 5.104006882925285, Z: 0}
+		expectedOri = &spatialmath.R4AA{
+			RX:    0.9864461301028694,
+			RY:    0.16360809262540335,
+			RZ:    0.012506975355798564,
+			Theta: 0.02398663944371901,
+		}
+	}
```
[the remark this answers is not in the document]
Comment (jeremyrhyde) [reply integration_test.go:83]: Yep! Tested on both mac, arm64 canon docker container and in github CI

File: sensorprocess/sensorprocess.go (hunk 6 above)
Comment (jeremyrhyde) [reply sensorprocess/sensorprocess.go:182]: Not too much as this start time is just used to calculate time elapsed which is independent of timezone. Other instances of time not being cast to UTC did cause issues so I decided to use it everywhere. However, I am happy to revert if desired.

File: sensors/test_deps.go [hunk 8]
```diff
@@ -18,14 +19,12 @@ import (
...
-	TestTime = "2006-01-02T15:04:05.9999Z"
-	// BadTime can be used to represent something that should cause an error while parsing it as a time.
-	BadTime = "NOT A TIME"
-)
+// BadTime can be used to represent something that should cause an error while parsing it as a time.
+const BadTime = "NOT A TIME"
 
 var (
+	// TestTimestamp can be used to test specific timestamps provided by a replay sensor.
+	TestTimestamp = time.Now().UTC().Format("2006-01-02T15:04:05.999999Z")
```
[the remark this answers is not in the document]
Comment (jeremyrhyde) [reply sensors/test_deps.go:27]: With all the mock integration sensors and other test functions using `time.Now()` as a base (adding and subtracting `time.Duration`s from it as desired). I wished to ensure a consistent source of testing that did not rely on an increasing out of data value

File: sensorprocess/sensorprocess.go (hunk 5 above)
Comment (jeremyrhyde) [reply sensorprocess/sensorprocess.go]: Done

File: sensorprocess/sensorprocess.go (hunk 4 above)
Comment (jeremyrhyde) [reply sensorprocess/sensorprocess.go]: Done

File: integration_test.go [hunk 9]
```diff
@@ -196,194 +247,124 @@ func testHelperCartographer(
...
+			subAlgo:     viamcartographer.Dim2d,
+		},
+		{
+			description: "replay sensor updating mode 2D",
+			replay:      true,
+			imuEnabled:  false,
+			mode:        cartofacade.UpdatingMode,
+			subAlgo:     viamcartographer.Dim2d,
+		},
+		// Live + imu sensor
```
[the remark this answers is not in the document]
Comment (kim-mishra) [reply integration_test.go:306]: cool, could you update the comment to reflect the correct tests?

**Review by kim-mishra — APPROVED** [review]
LGTM! ty for the refactor it looks a lot cleaner to me

File: testhelper/integrationtesthelper.go [hunk 10]
```diff
@@ -0,0 +1,331 @@
...
+
+	var i uint64
+	closed := false
+	ts := &s.TimedIMUSensorMock{}
+	ts.TimedIMUSensorReadingFunc = func(ctx context.Context) (s.TimedIMUSensorReadingResponse, error) {
+		defer timeTracker.Mu.Unlock()
+		/*
+			Holds the process until for all necessary lidar data has been sent to cartographer. Is always
+			true the first iteration. This combined with manual definition of timestamps allow for consistent
+			results.
```
Comment (kkufieta) [inline testhelper/integrationtesthelper.go:179]:
```suggestion
			Holds the process until all necessary lidar data has been sent to cartographer. Is always
			true in the first iteration. This and the manual definition of timestamps allow for consistent
			results.
```

File: testhelper/integrationtesthelper.go [hunk 11]
```diff
@@ -0,0 +1,331 @@
...
+			timeTracker.Mu.Lock()
+			if i == 0 || timeTracker.ImuTime.Sub(timeTracker.NextLidarTime) < 0 {
+				time.Sleep(sensorDataIngestionWaitTime)
+				break
+			}
+			timeTracker.Mu.Unlock()
+		}
+
+		// Communicating all desired imu readings have been sent or to cartographer or if the last lidar reading has been sent by
+		// checking if LastLidarTime has been defined. If so, simulate endOfDataSet error.
```
Comment (kkufieta) [inline testhelper/integrationtesthelper.go:191]:
```suggestion
		// Communicate that all desired IMU readings have been sent to cartographer or if the last lidar reading has been sent by
		// checking if LastLidarTime has been defined. If so, simulate endOfDataSet error.
```

**Review by kkufieta — CHANGES_REQUESTED** [review]
First pass. My suggestions regarding changes to the comments are guesses at what might be meant by them. Feel free to adjust them if they're not correct.

File: integration_test.go [hunk 12]
```diff
@@ -133,53 +150,87 @@ func testHelperCartographer(
...
-	// We will check both channels once an accurate mock dataset has been gathered,
-	// see https://viam.atlassian.net/browse/RSDK-4495
+	logger.Debug("lidar data finished")
 
+	// Wait for IMU sensor process to complete (optional)
+	if useIMU && !utils.SelectContextOrWaitChan(ctx, imuDone) {
+		t.Logf("test duration %dms", time.Since(start).Milliseconds())
+		test.That(t, errors.New("test timeout"), test.ShouldBeNil)
+	}
+	logger.Debug("imu data finished")
```
[the remark this answers is not in the document]
Comment (jeremyrhyde) [reply integration_test.go]: outdated based on more recent changes. Let me know if you agree

File: integration_test.go [hunk 13]
```diff
@@ -133,53 +150,87 @@ func testHelperCartographer(
...
-	// wait till all sensor readings have been read
+
+	// Wait for lidar sensor process to complete (required)
 	if !utils.SelectContextOrWaitChan(ctx, lidarDone) {
 		t.Logf("test duration %dms", time.Since(start).Milliseconds())
 		test.That(t, errors.New("test timeout"), test.ShouldBeNil)
 	}
-	// We will check both channels once an accurate mock dataset has been gathered,
-	// see https://viam.atlassian.net/browse/RSDK-4495
+	logger.Debug("lidar data finished")
```
[the remark this answers is not in the document]
Comment (jeremyrhyde) [reply integration_test.go]: outdated based on more recent changes. Let me know if you agree

File: testhelper/integrationtesthelper.go [hunk 14]
```diff
@@ -0,0 +1,331 @@
...
+
+	return ts, nil
+}
+
+// IntegrationTimedIMUSensor returns a mock timed IMU sensor.
+// When the mock is called, it returns the next mock IMU readings, with the
+// ReadingTime incremented by the sensorReadingInterval.
+// The Replay sensor field of the mock readings will match the replay parameter.
+// When the end of the mock IMU readings is reached, the done channel
+// is written to once so the caller can detect all IMU readings have been emitted
```
[the remark this answers is not in the document]
Comment (jeremyrhyde) [reply testhelper/integrationtesthelper.go:146]: Added when earlier in the line as it made more sense to me

**Review by kkufieta — APPROVED** [review]
LGTM!
